pkg/pgmodel/ingestor/watcher.go

```diff
@@ -0,0 +1,107 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package ingestor
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/timescale/promscale/pkg/log"
+)
+
+const (
+	// reportThreshold is a value above which the warn would be logged. This value is
+	// the value of the ratio between incoming batches vs outgoing batches.
+	reportRatioThreshold = 3
+	refreshIteration     = 10 // This means that values of watcher.incomingBatches and watcher.outgoingBatches gets renewed every (refreshIteration * checkThroughputInterval).
+	checkRatioInterval   = time.Minute
+
+	// Duration warnings constants.
+	reportDurationThreshold        = time.Minute
+	checkIngestionDurationInterval = time.Minute
+)
+
+// throughtputWatcher is a light weight samples batch watching type, that serves as a watching
+// routine and keeps the track of incoming write batches. It keeps a ratio of
+// incoming samples vs outgoing samples and warn about low throughput to the user
+// which might be due to external causes like network latency, resources allocated, etc.
+//
+// We watch the batches than the samples in total, since watching batching is fast
+// and does the same purpose of watching samples, but on a higher-level.
+type throughtputWatcher struct {
+	// Warn based on ratio.
+	incomingBatches uint64
+	outgoingBatches uint64
+
+	// Warn based on ingestion duration.
+	shouldReportLastDuration atomic.Value
+	lastIngestionDuration    atomic.Value
+}
+
+var watcher *throughtputWatcher
+
+func init() {
+	watcher = &throughtputWatcher{
+		incomingBatches: 0,
+		outgoingBatches: 0,
+	}
+	watcher.shouldReportLastDuration.Store(false)
+	go watcher.watch()
+	go watcher.watchTime()
```

Comment:
These two goroutines need to shut down cleanly. See https://github.com/timescale/promscale/commit/dad1c6f660e30c602bd64e0a2186a0ec3f3f8434 for an example